File: backend/models/audio.py
"""
Audio Models - Pydantic & SQLAlchemy
Production-Ready Audio Generation Models
"""

from pydantic import BaseModel, Field
from typing import Optional, Literal, List
from datetime import datetime
from enum import Enum

# ============================================
# Enums
# ============================================

class TTSProvider(str, Enum):
    """TTS Provider types"""
    OPENAI = "openai"
    SPEECHIFY = "speechify"
    GOOGLE = "google"
    MICROSOFT = "microsoft"
    ELEVENLABS = "elevenlabs"

class AudioFormat(str, Enum):
    """Audio output formats"""
    MP3 = "mp3"
    WAV = "wav"
    OGG = "ogg"
    AAC = "aac"

class AudioStatus(str, Enum):
    """Audio generation status"""
    PENDING = "pending"
    PROCESSING = "processing"
    COMPLETED = "completed"
    FAILED = "failed"

# ============================================
# Pydantic Models (API DTOs)
# ============================================

class AudioGenerateRequest(BaseModel):
    """Audio generation request"""
    text: str = Field(..., min_length=1, max_length=100000)
    provider: TTSProvider
    voice: str
    speed: float = Field(default=1.0, ge=0.25, le=4.0)
    format: AudioFormat = AudioFormat.MP3
    
class AudioGenerateResponse(BaseModel):
    """Audio generation response"""
    audio_id: str
    status: AudioStatus
    audio_url: Optional[str] = None
    duration_seconds: Optional[float] = None
    file_size_bytes: Optional[int] = None
    character_count: int
    cost_usd: float
    provider: TTSProvider
    voice: str
    
    class Config:
        from_attributes = True

class VoiceInfo(BaseModel):
    """Voice information"""
    id: str
    name: str
    language: str
    gender: Optional[str] = None
    preview_url: Optional[str] = None
    description: Optional[str] = None
    is_premium: Optional[bool] = False
    price_per_token: Optional[float] = None

class ProviderInfo(BaseModel):
    """TTS Provider information"""
    provider: TTSProvider
    name: str
    available: bool
    voices: List[VoiceInfo]
    cost_per_character: float
    max_characters: int

# ============================================
# SQLAlchemy ORM Models
# ============================================

from sqlalchemy import Column, String, DateTime, Enum as SQLEnum, Float, Integer, Text, Boolean
from sqlalchemy.ext.declarative import declarative_base
from sqlalchemy.sql import func
import uuid

Base = declarative_base()

class AudioGeneration(Base):
    """Audio generation history table"""
    __tablename__ = "audio_generations"
    
    id = Column(String(36), primary_key=True, default=lambda: str(uuid.uuid4()))
    user_id = Column(String(36), nullable=False, index=True)
    
    # Input
    text = Column(Text, nullable=False)
    character_count = Column(Integer, nullable=False)
    
    # Configuration
    provider = Column(SQLEnum(TTSProvider), nullable=False)
    voice = Column(String(100), nullable=False)
    speed = Column(Float, default=1.0, nullable=False)
    format = Column(SQLEnum(AudioFormat), default=AudioFormat.MP3, nullable=False)
    
    # Output
    status = Column(SQLEnum(AudioStatus), default=AudioStatus.PENDING, nullable=False)
    audio_url = Column(String(500), nullable=True)
    duration_seconds = Column(Float, nullable=True)
    file_size_bytes = Column(Integer, nullable=True)
    
    # Costs
    cost_usd = Column(Float, default=0.0, nullable=False)
    
    # Error handling
    error_message = Column(Text, nullable=True)
    
    # Timestamps
    created_at = Column(DateTime(timezone=True), server_default=func.now(), nullable=False)
    completed_at = Column(DateTime(timezone=True), nullable=True)
    
    def __repr__(self):
        return f"<AudioGeneration(id={self.id}, provider={self.provider}, status={self.status})>"

File: frontend/src/types/audio.ts
/**
 * Audio/TTS Types - Synced with Backend Models
 * Backend: models/audio.py
 * 
 * CRITICAL: These types MUST match Backend DTOs exactly!
 * Last synced: 2025-10-06
 */

// ============================================
// Enums (matching Backend)
// ============================================

export enum TTSProvider {
  OPENAI = 'openai',
  SPEECHIFY = 'speechify',
  GOOGLE = 'google',
  MICROSOFT = 'microsoft',
  ELEVENLABS = 'elevenlabs'
}

export enum AudioFormat {
  MP3 = 'mp3',
  WAV = 'wav',
  OGG = 'ogg',
  AAC = 'aac'
}

export enum AudioStatus {
  PENDING = 'pending',
  PROCESSING = 'processing',
  COMPLETED = 'completed',
  FAILED = 'failed'
}

// ============================================
// Audio Types (matching Backend DTOs)
// ============================================

export interface AudioGenerateRequest {
  text: string
  provider: TTSProvider
  voice: string
  speed?: number // 0.25 to 4.0, default 1.0
  format?: AudioFormat // default MP3
}

export interface AudioGenerateResponse {
  audioId: string
  status: AudioStatus
  audioUrl: string | null
  durationSeconds: number | null
  fileSizeBytes: number | null
  characterCount: number
  costUsd: number
  provider: TTSProvider
  voice: string
}

export interface VoiceInfo {
  id: string
  name: string
  language: string
  gender: string | null
  previewUrl: string | null
  description: string | null
  isPremium?: boolean
  pricePerToken?: number
}

export interface ProviderInfo {
  provider: TTSProvider
  name: string
  available: boolean
  voices: VoiceInfo[]
  costPerCharacter: number
  maxCharacters: number
}

// ============================================
// Voice Mappings (Helper)
// ============================================

export const VOICE_MAPPINGS: Record<TTSProvider, string[]> = {
  // OpenAI TTS - 11 voices (Updated 2025)
  [TTSProvider.OPENAI]: ['alloy', 'ash', 'ballad', 'coral', 'echo', 'fable', 'onyx', 'nova', 'sage', 'shimmer', 'verse'],

  // Speechify - System voices (verified 2025)
  [TTSProvider.SPEECHIFY]: ['henry', 'mia', 'george', 'jessica', 'snoop'],

  // ElevenLabs - Fallback voices (100+ loaded dynamically via API)
  // These are only used if API call fails
  [TTSProvider.ELEVENLABS]: [
    '21m00Tcm4TlvDq8ikWAM',  // Rachel - Fallback only
    'AZnzlk1XvdvUeBnXmlld',  // Domi - Fallback only
    'EXAVITQu4vr4xnSDxMaL',  // Bella - Fallback only
    'ErXwobaYiN019PkySvjV',  // Antoni - Fallback only
    'MF3mGyEYCl7XYWbV9V6O',  // Elli - Fallback only
    'TxGEqnHWrfWFTfGW9XjX',  // Josh - Fallback only
    'VR6AewLTigWG4xSOukaG',  // Arnold - Fallback only
    'pNInz6obpgDQGcFmaJgB',  // Adam - Fallback only
    'yoZ06aMxZJJ28mfd3POQ'   // Sam - Fallback only
  ],

  // Google TTS (gTTS - FREE!)
  [TTSProvider.GOOGLE]: [
    'de',    // German
    'en',    // English
    'es',    // Spanish
    'fr',    // French
    'it',    // Italian
    'pt',    // Portuguese
    'nl',    // Dutch
    'pl'     // Polish
  ],

  // Microsoft Edge TTS (edge-tts - FREE, 559 voices!)
  [TTSProvider.MICROSOFT]: [
    'de-DE-KatjaNeural',   // German Female
    'de-DE-ConradNeural',  // German Male
    'en-US-AriaNeural',    // English US Female
    'en-US-GuyNeural',     // English US Male
    'en-GB-SoniaNeural',   // English UK Female
    'en-GB-RyanNeural',    // English UK Male
    'es-ES-ElviraNeural',  // Spanish Female
    'fr-FR-DeniseNeural'   // French Female
  ]
}

// ============================================
// Type Guards
// ============================================

export const isAudioGenerateResponse = (obj: any): obj is AudioGenerateResponse => {
  return (
    typeof obj === 'object' &&
    typeof obj.audioId === 'string' &&
    Object.values(AudioStatus).includes(obj.status) &&
    typeof obj.characterCount === 'number'
  )
}

export const isProviderInfo = (obj: any): obj is ProviderInfo => {
  return (
    typeof obj === 'object' &&
    Object.values(TTSProvider).includes(obj.provider) &&
    typeof obj.available === 'boolean' &&
    Array.isArray(obj.voices)
  )
}
